File: c++/src/galTiling.cc
#include <iostream>
#include <iomanip>
#include <stdio.h>
#include <fstream>
#include <vector>
#include <map>
#include <cmath>
#include <time.h>
#include <boost/filesystem.hpp>
#include <boost/format.hpp>
#include <boost/regex.hpp>
#include <boost/program_options.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/algorithm/string.hpp>
#include "location.h"
using namespace std;
using boost::format;
namespace fs = boost::filesystem;
namespace po = boost::program_options;

struct GalRecord {
  string id;
  string tId;
  uint32_t tBeg;
  uint32_t tEnd;
  uint32_t tSize;
  string tSrd;
  
  string qId;
  uint32_t qBeg;
  uint32_t qEnd;
  uint32_t qSize;
  string qSrd;
  
  uint32_t ali;
  uint32_t mat;
  uint32_t mis;
  uint32_t qN;
  uint32_t tN;
  float ident;
  float score;

  LocVec qLoc;
  LocVec tLoc;
};
typedef vector<GalRecord> GalRecords;

void check_gal_record(const GalRecord& br) {
  if(br.qBeg > br.qEnd)
    cerr << format("%s: qBeg[%d] > qEnd[%d]\n") % br.qId % br.qBeg % br.qEnd;
  if(br.tBeg > br.tEnd)
    cerr << format("%s: tBeg[%d] > tEnd[%d]\n") % br.tId % br.tBeg % br.tEnd;
  if(br.score <= 0)
    cerr << format("%s:%d-%d score:%g\n") % br.qId % br.qBeg % br.qEnd % br.score;
  if(br.qLoc.size() != br.tLoc.size())
    cerr << format("unequal blocks: %s:%d-%d\n") % br.qId % br.qBeg % br.qEnd; 
  if(locVecLen(br.qLoc) != locVecLen(br.tLoc))
    cerr << format("unequal alnlen: %s:%d-%d\n") % br.qId % br.qBeg % br.qEnd; 
}
GalRecord make_gal_record(const vector<string>& ss) {   
  using boost::lexical_cast;
  
  GalRecord br;
  br.id = ss[0];

  br.tId = ss[1];
  br.tBeg = lexical_cast<uint32_t>(ss[2]);
  br.tEnd = lexical_cast<uint32_t>(ss[3]);
  br.tSrd = ss[4];
  br.tSize = lexical_cast<uint32_t>(ss[5]);
  
  br.qId = ss[6];
  br.qBeg = lexical_cast<uint32_t>(ss[7]);
  br.qEnd = lexical_cast<uint32_t>(ss[8]);
  br.qSrd = ss[9];
  br.qSize = lexical_cast<uint32_t>(ss[10]);
   
  br.ali = ss[11].empty() ? 0 : lexical_cast<uint32_t>(ss[11]);
  br.mat = ss[12].empty() ? 0 : lexical_cast<uint32_t>(ss[12]);
  br.mis = ss[13].empty() ? 0 : lexical_cast<uint32_t>(ss[13]);
  br.qN  = ss[14].empty() ? 0 : lexical_cast<uint32_t>(ss[14]);
  br.tN  = ss[15].empty() ? 0 : lexical_cast<uint32_t>(ss[15]);
  br.ident = ss[16].empty() ? 0 : lexical_cast<float>(ss[16]);
  br.score = ss[17].empty() ? lexical_cast<float>(br.mat) :  lexical_cast<float>(ss[17]);

  br.qLoc = locStr2Vec(ss[18]);
  br.tLoc = locStr2Vec(ss[19]);

  check_gal_record(br);
  return br;
}

void gal_tiling(const GalRecords& grs, string& qId, ofstream& fho, const unsigned& len_min) {
  LocVec lv1;
  int i = 0;
  for(GalRecords::const_iterator it = grs.begin(); it != grs.end(); it++) {
    Location loc;
    loc.beg = it->qBeg;
    loc.end = it->qEnd;
    loc.score = it->score;
    loc.idxs.insert( i++ );
    lv1.push_back(loc);
  }

  LocVec lv2 = tiling(lv1, true);
  for(LocVec::const_iterator it = lv2.begin(); it != lv2.end(); it++) {
    Location loc = *it;
    uint32_t qBeg(loc.beg), qEnd(loc.end);
    uint32_t qLen = qEnd - qBeg + 1;
    if(qLen < len_min) continue;
    
    GalRecord gr = grs[ loc.idx_ext ];
    string id(gr.id), qSrd("+");
    string tSrd = gr.qSrd == gr.tSrd ? "+" : "-";
    uint32_t tSize(gr.tSize), qSize(gr.qSize);
    float ident(gr.ident), score(gr.score);
     
    string tId = gr.tId;
    uint32_t rqb(qBeg-gr.qBeg+1), rqe(qEnd-gr.qEnd+1);
    Location rqloc;
    rqloc.beg = rqb;
    rqloc.end = rqe;
    LocVec rqLoc;
    rqLoc.push_back(rqloc);
    
    LocVec qLoc = loc_intersect(gr.qLoc, rqLoc);
    rqb = qLoc.begin()->beg;
    rqe = qLoc.rbegin()->end;
    qBeg = gr.qBeg + rqb - 1;
    qEnd = gr.qEnd + rqe - 1;
    
    uint32_t rtb = coordTransform(rqb, gr.qLoc, "+", gr.tLoc, "+"); //need fill
    uint32_t rte = coordTransform(rqe, gr.qLoc, "+", gr.tLoc, "+");
    Location rtloc;
    rtloc.beg = rtb;
    rtloc.end = rte;
    LocVec rtLoc;
    rtLoc.push_back(rtloc);

    LocVec tLoc = loc_intersect(gr.qLoc, rtLoc);
    uint32_t tBeg = tSrd=="+" ? gr.tBeg+rtb-1 : gr.tEnd-rte+1;
    uint32_t tEnd = tSrd=="+" ? gr.tBeg+rte-1 : gr.tEnd-rtb+1;
    fho << format("%s\t%s\t%d\t%d\t%s\t%d\t%s\t%d\t%d\t%s\t%d \
      \t\t\t\t\t\t%g\t%g\t%s\t%s\n") 
      % id % tId % tBeg % tEnd % tSrd % tSize
           % qId % qBeg % qEnd % qSrd % qSize 
      % ident % score % locVec2Str(qLoc) % locVec2Str(tLoc);
  }
  cout << qId << endl;
}

int main( int argc, char* argv[] ) {
  string fi, fo;
  unsigned len_min;
  clock_t time1 = clock();
  po::options_description cmdOpts("Allowed options");
  cmdOpts.add_options()
    ("help,h", "produce help message")
    ("in,i", po::value<string>(&fi), "input file")
    ("out,o", po::value<string>(&fo), "output file")
    ("min,m", po::value<unsigned>(&len_min)->default_value(1), "mininum tiling length")
  ;
  po::variables_map vm;
  po::store(po::parse_command_line(argc, argv, cmdOpts), vm);
  po::notify(vm);
  if(vm.count("help") || !vm.count("in") || !vm.count("out")) {
    cout << cmdOpts << endl;
    return 1;
  }

  ifstream fhi(fi.c_str());
  if(!fhi.is_open()) { cout << format("cannot read: %s\n") % fi; return false; }
  ofstream fho(fo.c_str());
  if(!fho.is_open()) { cout << format("cannot write: %s\n") % fo; return false; }
  fho << "id\ttId\ttBeg\ttEnd\ttSrd\ttSize"
    << "\tqId\tqBeg\tqEnd\tqSrd\tqSize"
    << "\tali\tmat\tmis\tqN\ttN\tident\tscore\tqLoc\ttLoc" << endl;
  
  GalRecords grs;
  string qId_p = "";
  string line;
  while(fhi.good()) {
    getline(fhi, line);
    if(line.length() == 0) continue;
    boost::erase_all(line, " ");
    
    vector<string> ss;
    boost::split(ss, line, boost::is_any_of("\t"));
    if(ss[0] == "id") continue;
    
    GalRecord gr = make_gal_record(ss);
    if(gr.qId == qId_p) {
      grs.push_back(gr);
    } else {
      if(qId_p != "") {
        gal_tiling(grs, qId_p, fho, len_min);
      }
      grs.clear();
      grs.push_back(gr);
      qId_p = gr.qId;
    }
  }
  if(grs.size() > 0) {
    gal_tiling(grs, qId_p, fho, len_min);
  }

  cout << right << setw(60) << format("(running time: %.01f minutes)\n") % ( (double)(clock() - time1) / ((double)CLOCKS_PER_SEC * 60) );
  return EXIT_SUCCESS;
}

File: perl/gal.tiling.pl
#!/usr/bin/perl -w
#
# POD documentation
#---------------------------------------------------------------------------
=pod BEGIN
  
=head1 NAME
  
  gal.tiling.pl - tiles a Gal file

=head1 SYNOPSIS
  
  galtiling.pl [-help] [-in input-file] [-out output-file]

  Options:
    -h (--help)   brief help message
    -i (--in)     input (Gal) file
    -o (--out)    output (Gal) file
    -m (--min)    minimum tiling length (default: 10)

=cut
  
#### END of POD documentation.
#---------------------------------------------------------------------------

use strict;
use FindBin;
use lib "$FindBin::Bin";
use Getopt::Long;
use Pod::Usage;
use Common;
use Gal;
use Location;
use Data::Dumper;
use List::Util qw/min max sum/;

my ($fi, $fo) = ('') x 2;
my $minlen = 10;
my $help_flag;

#--------------------------------- MAIN -----------------------------------#
GetOptions(
  "help|h"   => \$help_flag,
  "in|i=s"  => \$fi,
  "out|o=s" => \$fo,
  "min|m=i" => \$minlen,
) or pod2usage(2);
pod2usage(1) if $help_flag;
pod2usage(2) if !$fi || !$fo;

runCmd("tail -n +2 $fi | sort -k7,7 -k8,8n -o $fi.tmp");

my ($fhi, $fho);
open ($fhi, "<$fi.tmp") || die "cannot read $fi.tmp\n";
open ($fho, ">$fo") || die "cannot write $fo\n";
print $fho join("\t", @HEAD_GAL)."\n";

my @lines;
my $cqid = '';
while(<$fhi>) {
  chomp;
  my ($id, $tId, $tBeg, $tEnd, $tSrd, $tSize, 
    $qId, $qBeg, $qEnd, $qSrd, $qSize,
    $lev, $ali, $mat, $mis, $qN, $tN, $ident, $score, $tls, $qls) = split "\t";
  my ($tl, $ql) = (locStr2Ary($tls), locStr2Ary($qls));
  my $line = [$id, $tId, $tBeg, $tEnd, $tSrd, $tSize,
      $qId, $qBeg, $qEnd, $qSrd, $qSize, $score, $tl, $ql];
  if($qId eq $cqid) {
    push @lines, $line;
  } elsif($cqid ne '') {
    gal_tile_one(\@lines, $minlen, $fho);
    backOneLine($fhi);
#    printf "%s: %d\n", $cqid, scalar(@lines);
    @lines = ();    
    $cqid = $qId;
  } else {
    push @lines, $line; 
    $cqid = $qId;
  }
}
gal_tile_one(\@lines, $minlen, $fho) if @lines > 0;

close $fhi;
close $fho;
runCmd("rm $fi.tmp");


sub gal_tile_one {
  my ($lines, $minlen, $fho) = @_;
  my @locs = map {[$_->[7], $_->[8]]} @$lines;
  my @scores = map {$_->[-3]} @$lines;
  my $ref = tiling(\@locs, \@scores, 2);
  for (@$ref) {
    my ($qb, $qe, $idx, $idxs) = @$_;
    next if $qe - $qb + 1 < $minlen;
    my ($id, $tId, $tBeg, $tEnd, $tSrd, $tSize, $qId, $qBeg, $qEnd, 
      $qSrd, $qSize, $score, $tl, $ql) = @{$lines->[$idx]};
    my ($rqb, $rqe) = $qSrd eq "+" ? ($qb-$qBeg+1, $qe-$qBeg+1) :
      ($qEnd-$qe+1, $qEnd-$qb+1);

    my ($idx1, $idx2);
    ($idx1, $rqb) = find_interval($rqb, $ql);
    ($idx2, $rqe) = find_interval($rqe, $ql);
    my $rtb = coordTransform($rqb, $ql, "+", $tl, "+");
    my $rte = coordTransform($rqe, $ql, "+", $tl, "+");
    my ($tb, $te) = $tSrd eq "+" ? ($tBeg+$rtb-1, $tBeg+$rte-1) :
      ($tEnd-$rte+1, $tEnd-$rtb+1);
    
    my ($ntl, $ntll) = posOvlp([[$rtb, $rte]], $tl);
    my ($nql, $nqll) = posOvlp([[$rqb, $rqe]], $ql);
    $ntl = [ map {[$_->[0]-$rtb+1, $_->[1]-$rtb+1]} @$ntl ];
    $nql = [ map {[$_->[0]-$rqb+1, $_->[1]-$rqb+1]} @$nql ];
    my ($ntls, $nqls) = (locAry2Str($ntl), locAry2Str($nql));
    my $ali = locAryLen($ntl);
    my $str = "[$qBeg-$qEnd] [$qb-$qe] [$rqb-$rqe] [$nqls]\n".
      "[$tBeg-$tEnd] [$tb-$te] [$rtb-$rte] [$ntls]\n";
    
    print $fho join("\t", $id, $tId, $tb, $te, $tSrd, $tSize,
      $qId, $qb, $qe, $qSrd, $qSize, 
      '', $ali, ('')x5, $score, $ntls, $nqls)."\n";
  }
}
